File: androidApp/src/main/java/com/nagyrobi144/dogify/android/MainViewModel.kt
package com.nagyrobi144.dogify.android

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.nagyrobi144.dogify.model.Breed
import com.nagyrobi144.dogify.repository.BreedsRepository
import com.nagyrobi144.dogify.usecase.FetchBreedsUseCase
import com.nagyrobi144.dogify.usecase.GetBreedsUseCase
import com.nagyrobi144.dogify.usecase.ToggleFavouriteStateUseCase
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch

class MainViewModel(
    breedsRepository: BreedsRepository,
    private val getBreeds: GetBreedsUseCase,
    private val fetchBreeds: FetchBreedsUseCase,
    private val onToggleFavouriteState: ToggleFavouriteStateUseCase
) : ViewModel() {

    private val _state = MutableStateFlow(State.LOADING)
    val state: StateFlow<State> = _state

    private val _isRefreshing = MutableStateFlow(false)
    val isRefreshing: StateFlow<Boolean> = _isRefreshing

    private val _events = MutableSharedFlow<Event>()
    val events: SharedFlow<Event> = _events

    private val _shouldFilterFavourites = MutableStateFlow(false)
    val shouldFilterFavourites: StateFlow<Boolean> = _shouldFilterFavourites

    val breeds =
        breedsRepository.breeds.combine(shouldFilterFavourites) { breeds, shouldFilterFavourites ->
            if (shouldFilterFavourites) {
                breeds.filter { it.isFavourite }
            } else {
                breeds
            }.also {
                _state.value = if (it.isEmpty()) State.EMPTY else State.NORMAL
            }
        }.stateIn(
            viewModelScope,
            SharingStarted.WhileSubscribed(),
            emptyList()
        )

    init {
        loadData()
    }

    private fun loadData(isForceRefresh: Boolean = false) {
        val getData: suspend () -> List<Breed> =
            { if (isForceRefresh) fetchBreeds.invoke() else getBreeds.invoke() }

        if (isForceRefresh) {
            _isRefreshing.value = true
        } else {
            _state.value = State.LOADING
        }

        viewModelScope.launch {
            _state.value = try {
                getData()
                State.NORMAL
            } catch (e: Exception) {
                State.ERROR
            }
            _isRefreshing.value = false
        }
    }

    fun refresh() {
        loadData(true)
    }

    fun onToggleFavouriteFilter() {
        _shouldFilterFavourites.value = !shouldFilterFavourites.value
    }

    fun onFavouriteTapped(breed: Breed) {
        viewModelScope.launch {
            try {
                onToggleFavouriteState(breed)
            } catch (e: Exception) {
                _events.emit(Event.Error)
            }
        }
    }

    enum class State {
        LOADING,
        NORMAL,
        ERROR,
        EMPTY
    }

    enum class Event {
        Error
    }
}

File: iosApp/iosApp/MainViewModel.swift
//
//  MainViewModel.swift
//  iosApp
//
//  Created by Robert Nagy on 21/07/2021.
//  Copyright © 2021 orgName. All rights reserved.
//

import Foundation
import shared
import KMPNativeCoroutinesRxSwift
import RxSwift
import Combine

class MainViewModel: ObservableObject {
    
    private let repository = BreedsRepository.init()
    private let getBreeds = GetBreedsUseCase.init()
    private let fetchBreeds = FetchBreedsUseCase.init()
    private let onToggleFavouriteState = ToggleFavouriteStateUseCase.init()
    
    @Published
    private(set) var state = State.LOADING
    
    @Published
    var shouldFilterFavourites = false
    
    @Published
    private(set) var filteredBreeds: [Breed] = []
    
    @Published
    private var breeds: [Breed] = []
    
    private let disposeBag = DisposeBag()
    
    init() {
       createObservable(for: repository.breedsNative).subscribe(onNext: { breeds in
            DispatchQueue.main.async {
                self.breeds = breeds
            }
        }).disposed(by: disposeBag)

        $breeds.combineLatest($shouldFilterFavourites, { breeds, shouldFilterFavourites -> [Breed] in
            var result: [Breed] = []
            if(shouldFilterFavourites){
                result.append(contentsOf: breeds.filter{ $0.isFavourite })
            } else {
                result.append(contentsOf: breeds)
            }
            if(result.isEmpty){
                self.state = State.EMPTY
            } else {
                self.state = State.NORMAL
            }
            return result
        }).assign(to: &$filteredBreeds)
        
        getData()
    }
    
    func getData(){
        state = State.LOADING
        
        createSingle(for: getBreeds.invokeNative()).subscribe(onSuccess: { _ in
            DispatchQueue.main.async {
                self.state = State.NORMAL
            }
        }, onFailure: { error in
            DispatchQueue.main.async {
                self.state = State.ERROR
            }
        }).disposed(by: disposeBag)
    }
    
    func fetchData() {
        state = State.LOADING
        
        createSingle(for: fetchBreeds.invokeNative()).subscribe(onSuccess: { _ in
            DispatchQueue.main.async {
                self.state = State.NORMAL
            }
        }, onFailure: { error in
            DispatchQueue.main.async {
                self.state = State.ERROR
            }
        }).disposed(by: disposeBag)
    }
    
    func onFavouriteTapped(breed: Breed){
        createSingle(for: onToggleFavouriteState.invokeNative(breed: breed)).subscribe(onFailure: { error in
           // We're going ignoring the failure, as it will be represented by the stream of breds
        }).disposed(by: disposeBag)
    }
    
    enum State {
        case NORMAL
        case LOADING
        case ERROR
        case EMPTY
    }
    
}
